### DISCUSSION TITLE
Logical data model

### DISCUSSION THREAD

**Discussion by david-i-berry** [wmo-im/tt-cdm#9] [General]
Thread for to discuss and agree on logical data model for the conceptual data model agreed at

https://github.com/wmo-im/tt-cdm/discussions/8

and meeting the requirements specified at 

https://github.com/wmo-im/tt-cdm/discussions/7

**Comment by david-i-berry** [comment]
A key component of the data model is the representation of observations and previous WMO data models for observations (e.g. METCE) have been based on and extended the ISO 19156 / OGC Observations and Measurements data model. This is also used in the WMO WIGOS metadata representation (WMDR) data model. 

The same approach is proposed here, except updating to the new OGC Observations, Measurements and Samples data model and standard. This standard is due to be published in November 2022 as an update to the ISO19156 standard and the draft specification is available from: https://github.com/opengeospatial/om-swg. The conceptual model can be expanded to include the OGC elements (noting the renaming of sensor -> observer from the diagram on the conceptual model discussion page):

```mermaid
erDiagram
   dataset ||--|{ observation: contains
   observation }|--|| observer: "made using sensor"
   observation }|--|| host: "located at station"
   observation ||--|{ processing: "has processing applied"
```

Note, some of the elements, such as observable property have been omitted for clarity.

Within the OGC OMS model only the observation class is prescribed in any detail. The other entities, such as host and observer, are left generic to enable widespread adoption and use in many applications, not just specific to Earth system data.  Within the OGC OMS model there are only two mandatory elements for the host and observer classes, their location and a URI link providing a link to the description of the entities. Similarly, discovery metadata is provided by way of a link rather than a prescription of the contents. The class diagram below expands the different elements shown.

```mermaid
classDiagram
class Observation{   
   +id : string or int
   +observationType: AbstractObservationType
   +parameter: list~NamedValue~
   +phenomenonTime : TM_Object
   +resultQuality : Any
   +resultTime : TM_Object
   +validTime : TM_Period    
}

class Observer{
    +id : string or int
    +link: URI
    +location: Geometry
}

class Host {
    +id : string or int
    +link: URI
    +location: Geometry
}

class Metadata {
    +id : string or int
    +link: URI
}

class NamedValue {
    +name: string
    +value: Any
    +link: URI
}

class ObservingProcedure {
    +id: string or int
    +link: URI
}

class ObservableProperty {
    +id: string or int
    +link: URI
}

class Result {
    +id : string or int
    +value: Any
    +uom: URI or string
}

class Feature {
    +id: string or int
    +link: URI
    +location: Geometry
}

class Geometry {
    +type: string
    +coordinates: list~numeric~
    +wkt: string
}

Observation --> Observer : +observer
Observation --> Host : +host
Observation --> ObservingProcedure : + observingProcedure
Observation --> ObservableProperty : +observedProperty
Observation --> Metadata: +metadata
Observation --> Observation: +relatedObservation
Observation --> Result : +result
Observation --> Feature: +proximateFeatureOfInterest
Observation --> Feature: +ultimateFeatureOfInterest
```

Recognizing that the WMO has existing data models for the observational and discovery metadata it is recommended that these are used. This leads to the following additional constraints:

- the link:URI attribute for the Host and Observer classes SHALL reference records that conform with the WMDR data model.
- the link:URI attribute for the Metadata class SHALL reference a record that conforms with the WCMP2 data model.

Similar recommendations can be made for the ObservingProcedure and ObservableProperty classes:

- the link:URI attribute for the ObservingProcedure SHOULD reference an ObservingMethodType defined as part of the WMDR data model and code lists.
- the link:URI attribute for the ObservableProperty SHOULD reference an ObservedVariable defined as part of the WMDR data model and code lists.

The weaker SHOULD is used in recognition that not all WMO Application Areas or Domains are fully represented within the WMDR.

Note: the `NamedValue` class has been extended to allow external linked to a controlled vocabulary.

**Reply by joergklausen** [reply]
The WMDR should (sic!) evolve in response to requirements of WMO Application Areas or Domains. WDMR shall (sic!) enable documentation of observations for all  WMO Application Areas or Domains - this has been decided at the outset. As a community, we must work towards improving WMDR where it currently does not meet user needs. The process is sometimes slower than people desire, but the ambition remains a core driving force for the TT-WIGOSMD. CRs can always be placed and will be dealt with.

**Comment by david-i-berry** [comment]
Noting the above, the observation, host, observer and metadata can be described using existing data models, leaving only the processing to be defined. A class to record this processing is proposed below:

```mermaid
classDiagram
class ProcessingLog{
    +id: string or int
    +type: string
    +item: URI
    +path: string
    +oldValue: any
    +newValue: any
    +changeDate: TM_Object
    +changeDescription: string
    +changeReason: URI
    +changedBy: User
}
```

The ´type´ attribute is intended to reference the type of object the log entry refers to, for example ´Observation´ or ´Result´. The ´item´ is used to identify which item is referred to using a URI whilst ´path´ identifies the attribute modified. As with the other classes, the reason for the change can be described using a controlled vocabulary linked by the ´changeReason´ field.

**Comment by david-i-berry** [comment]
A word example, using the TTL language is given below:

```ttl
@prefix rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xs:   <http://www.w3.org/2001/XMLSchema#> .
@prefix skos:  <http://www.w3.org/2004/02/skos/core#>.
@prefix owl:   <http://www.w3.org/2002/07/owl#>.
@prefix : <http://example.com/cdm/> .
@base <http://example.com/collections/> .

# Observations index / table
<observations/1> a :Observation;
                 :id "10476F0.1999.1.1.MAX_TEMPERATURE";
                 :observationType "http://www.opengis.net/def/observation-type/OGC/0/by-result-type/observation";
                 :parameter _:p0, _:p1, _:p2, _:p3, _:p4, _:p5, _:p6, _:p7;
                 :phenomenonTime "1999-01-01T00:00.00/1999-01-01T23:59.59";
                 :resultQuality "Good";
                 :resultTime "1999-01-02T00:00.00";
                 :validTime "nilReason";
                 :observer <observers/1>;
                 :host <hosts/1>;
                 :observingProcedure <codes/wmdr/ObservingMethodAtmosphere/_149>;
                 :observedProperty <codes/wmdr/ObservedVariableAtmosphere/_224>;
                 :metadata <datasets/1>;
                 :result _:r1;
                 :proximateFeature _:f1;
                 :ultimateFeature "nilReason";
                 :relatedObservations "nilReason".

# Example feature of interest
_:f1 a :Feature;
     :id "Some UUID";
     :link <nilReason>;
     :location _:g1.

# Example geometry item
_:g1 a :Geometry;
     :type "Point";
     :coordinates (-123.1612 49.7832 53.7);
     :wkt "POINT((-123.1612 49.7832 53.7), SRID=4326)".

# Example result object
_:r1 a :Result;
     :id "r1";
     :value 4.7;
     :uom <https://github.com/wmo-im/wmds/blob/master/wmdr/unit/degC.ttl>.

# Example named parameters
_:p0 a :Parameter;
     :name "reportId";
     :value "10476F0.1999.1.1";
     :link <nilReason>.

_:p1 a :Parameter;
    :name "wigos_station_identifier";
    :value "0-20000-0-71207";
     :link <nilReason>.

_:p2 a :Parameter;
    :name "station_or_site_name";
    :value "Squamish Airport";
     :link <nilReason>.

_:p3 a :Parameter;
     :name "sensor_height_above_local_surface";
     :value 2;
     :link <nilReason>.

_:p4 a :Parameter;
     :name "sensor_height_above_sea_level";
     :value 55.7;
     :link <terms/sensor_height_above_sea_level>.

_:p5 a :Parameter;
     :name "siting_classification";
     :value "class1";
     :link <https://github.com/wmo-im/wmds/blob/master/wmdr/Exposure/class1.ttl>.

_:p6 a :Parameter;
     :name "measurement_quality_classification";
     :value "B";
     :link <terms/measurement_quality_classification>.

_:p7 a :Parameter ;
     :name "report_type";
     :value "DAYCLI";
     :link <terms/report_type>.

# Example station / host
<hosts/1> a :Host;
          :id "0-20000-0-71207";
          :link <https://example.com/stations/0-20000-0-71207>;
          :location _:g1.

# Example instrument / observer item
<observers/1> a :Observer ;
            :id "1-12345-0-ABC123";
            :link <https://example.com/instruments/1-12345-0-ABC123>;
            :location _:g1.

# Example dataset item
<datasets/1> a :Dataset;
             :id "ca-climate-daily";
             :link "https://api.weather.gc.ca/collections/climate-daily/" .

# Example log items
<log/1> a :Log;
        :id 1;
        :type "Result";
        :item _:r1;
        :path "value";
        :oldValue 5.0;
        :newValue 4.7;
        :changeDate "2022-08-08T10:30Z";
        :changeDescription "Corrected for erroneous rounding based on comparison with logbook";
        :changeReason <nilReason>;
        :changedBy <users/1> .

<log/2> a :Log;
        :id 2;
        :type "Observation";
        :item <observations/1>;
        :path "resultQuality";
        :oldValue "NA";
        :newValue "Good";
        :changeDate "2022-08-08T11:00Z";
        :changeDescription "QC flag updated following manual inspection";
        :changeReason <nilReason>;
        :changedBy <users/1> .

# Example code list for observing procedure
<codes/wmdr/ObservingMethodAtmosphere/_149> a :ObservingProcedure;
                                            :id "proc1";
                                            :link <http://codes.wmo.int/wmdr/ObservingMethodAtmosphere/_149>.

# Example code list for observed properties
<codes/wmdr/ObservedVariableAtmosphere/_224> a :ObservedProperty ;
                                             :id "prop1";
                                             :link <http://codes.wmo.int/wmdr/ObservedVariableAtmosphere/_226> .
```

**Reply by peterwthorne** [reply]
We had raised on the call the concern about pointing to resources that might not be long-term sticky. There would be two risks here: the pointer becoming stale; and the pointer pointing to something other than intended if the external resource gets reorganised and identifiers /codes etc. reassigned. I'm not clear that there is a way to absolutely avoid this but perhaps some principles as to what is an acceptable resource to point to externally might at least mitigate.

**Reply by joergklausen** [reply]
I agree, reference to github is not a good idea. However, referencing the codes.wmo.int/wmdr should be robust enough. This should be changed in _:r1 and _:p5 above. Where such references do not currently exist on codes.wmo.int/wmdr, they should be added by way of a CR.

**Comment by david-i-berry** [comment]
As part of the OpenCDMS project we are now starting to implement a data model with test data. The model we are using is shown below.

```mermaid
classDiagram
class Observation{   
   +id : string or int
   +location: Geometry
   +observationType: AbstractObservationType
   +phenomenonTime : TM_Object
   +result: Result
   +resultTime : TM_Object
   +resultQuality : Any
   +validTime : TM_Period    
   +parameter: list~NamedValue~
   +version: int
   +changeDate: TM_Instance
   +comments: string
}

class ObservationType{
   +id: string or int
   +name: string
   +description: string
   +link: URI
}

class RecordStatus{
   +id: string or int
   +name: string
   +description: string
}

class Users{
   +id: string or int
   +name: string
}

class Observer{
    +id: string or int
    +name: string
    +description: string
    +link: URI
    +location: Geometry
}

class Host {
    +id : string or int
    +name: string
    +description: string
    +link: URI
    +location: Geometry
    +elevation: numeric
    +wsi: string
    +facility_type: URI
    +date_established: TM_Instance
    +wmo_region: URI
   +territory: URI
   +validPeriod: TM_Period
   +version: int
   +change_date
   +comments: string
}

class Metadata {
    +id : string or int
    +name: string
    +link: URI
}

class NamedValue {
    +name: string
    +value: Any
    +link: URI
}

class ObservingProcedure {
   +id: string or int
   +name: string
   +description: string
   +link: URI
}

class ObservableProperty {
    +id: string or int
    +shortName: string
    +standardName: string
    +units: string
    +description: string
    +link: URI
}

class Result {
    +value: Any
    +uom: URI or string
    +description: string
}

class Feature {
    +id: string or int
    +location: Geometry
    +name: string
    +description: string
    +link: URI
}

class Geometry {
    +type: string
    +coordinates: list~numeric~
    +wkt: string
}

Observation --> Observer : +observer
Observation --> Host : +host
Observation --> ObservingProcedure : + observingProcedure
Observation --> ObservableProperty : +observedProperty
Observation --> Metadata: +metadata
Observation --> Observation: +relatedObservation
Observation --> Feature: +proximateFeatureOfInterest
Observation --> Feature: +ultimateFeatureOfInterest
Observation --> RecordStatus: +status
Observation --> Users: +user
Observation --> ObservationType: +observationType
Host--> RecordStatus: +status
Host --> Users: +user
Feature --> Feature: +parentFeature

```

This is similar to that shown above but with an expansion of the host to include the minimal information we need at this stage.

**Reply by david-i-berry** [reply]
Noting Peters above comment (from 26th August) we will need to add rules around the linking / use of controlled vocabularies and code tables.

**Comment by david-i-berry** [comment]
Comments from Denis (truncated)

> Quickly, what would be the first main concerns for a climatological point of view are: 
> 
> * How the traceability is managed with the model? e.g. data modifications, Associated quality modifications, provenance modifications, etc. 
> * How we plan to assign qualities to data with the model?

**Reply by david-i-berry** [reply]
Following discussions last year, as part of the TT-CDM, but also separately with Jörg and Kathy as part of a discussion on the OGC OMS model, I got the impression that there was a preference to duplicate and version a full record when it was changed rather than just the changed fields. As such , I've added the following fields to the observation:

- user
- version
- status
- change date
- comments

These can either be part of the named value list (parameter field) or higher up in the observation object as in the latest version. There are pros and cons to both, putting them in the named value list may be more consistent with OMS.

**Reply by david-i-berry** [reply]
Quality flags can be assigned to the resultQuality field. This is of type any so can also be a named value list to allow the reporting of multiple QC flags.

**Reply by joergklausen** [reply]
> Following discussions last year, as part of the TT-CDM, but also separately with Jörg and Kathy as part of a discussion on the OGC OMS model, I got the impression that there was a preference to duplicate and version a full record when it was changed rather than just the changed fields. As such , I've added the following fields to the observation:
> 
> * user
> * version
> * status
> * change date
> * comments
> 
> These can either be part of the named value list (parameter field) or higher up in the observation object as in the latest version. There are pros and cons to both, putting them in the named value list may be more consistent with OMS.

IMHO, within the limits of the model (i.e., what elements are mandatory to generate a valid record), partial records should be supported. I believe, this makes it easier for the user and perhaps even lighter for the systems that have to deal with a record. The meta information on user, version etc of a record are part of a header section in WMDR, and this approach has worked well. I think it can be recommended here as well.

**Reply by david-i-berry** [reply]
> The meta information on user, version etc of a record are part of a header section in WMDR, and this approach has worked well. I think it can be recommended here as well.

Is this a part of the data model or the way the WMDR has been implemented as part of OSCAR? The information available in the header is fairly basic, version of the XSD, last change date and owner but no change log or version of the data itself.

Also in looking through the UML diagram and the schema specification document [https://schemas.wmo.int/wmdr/1.0RC9/documentation/WMDR_ModelAndSchemaSpecification.pdf](https://schemas.wmo.int/wmdr/1.0RC9/documentation/WMDR_ModelAndSchemaSpecification.pdf) it appears that the WIGOS Identifiers do not form part of the data model (but are included in the schema). I would have expected to see the WIGOS identifier to appear in the AbstractEnvironmentalMonitoringFacility class.

**Comment by david-i-berry** [comment]
Additional comment from Dennis:

> Hi,
>  
> Just a thought for the model,  about one of the first tasks for a CDMS: to ensure data completeness. To be able to check the completeness of a system you need: 
> 
> (1) to know which data is expected and at which frequency,  (this may be different from the capacity of the full data acquisition process of the NMHS, e.g. you may be able to acquire minute data from your network but you expect only 10 minute data to be acquired);
> (2) to know which data is present at any time (the reality).
> 
> The difference between the two allows you to do the Monitoring and the checking for the completeness. So, the Date/Time of each insert of a data value should be known in the CDMS.
> 
> Denis.
> 
> 
> Denis Stuber

**Reply by david-i-berry** [reply]
The time an observation became available is given by the resultTime attribute and so is known in the CDMS. We also have the changeDate which gives the last date / time when an observation was modified.

**Reply by joergklausen** [reply]
I agree.

**Reply by ET-DRC** [reply]
Thanks for the comments.
Is the resultTime the availability time of the data (over the data flow, or for the AWS, or over the data Concentrator, etc.) or the time of insertion into the database ?

Joerg, ET-DRC has currently a discussion about the concept of deployment in OSCAR. I attach the doc of the discussion to this message. The Siting Classification and the Measurement quality classification traceability are raised as well as the need to manage multiple instruments for the same variable. Do you agree with this paper ?
Denis.
[Deployment.pdf](https://github.com/wmo-im/tt-cdm/files/10455357/Deployment.pdf)

**Reply by david-i-berry** [reply]
From the OMS standard:

![image](https://user-images.githubusercontent.com/29403616/213410366-9470abec-8ef9-48e4-8c24-b1e0877be734.png)

Example 4 probably covers our case (but may have implications for the change_date). we can also store additional dates and times in the parameter attribute.

One of the big tasks once we can agree on the data model will be the governance of the controlled vocabularies used, for example the names used in the named value pairs from the parameter attribute.

**Reply by scottylad501** [reply]
This is a very important point that Denis raises and something Belize has thought about extensively in developing the SURFACE CDMS....one issue raised during development was, will NMHSs state for each station the expected frequency and if for some reason the frequency changes will the NMHS update the metadata to reflect the change. As Denis rightly states this has implication for the data inventory(thats what we call it) to determine if there is missing data etc as well as data aggregations i.e. creating hourly/daily data from the "raw" sub hourly data.  What we decided to do was to let the CDMS track the raw data time stamps to determine the frequency of data, so it compares the one time stamp to the previous time stamp to determine the frequency...so it knows if its 1min data, 5min data, 10min data etc  However for the data inventory instead of having an expected number we know for example that 5 min data which is the default frequency in Belize we know its 288 values and some of our stations store in 15mins which is 96 values.

**Reply by david-i-berry** [reply]
In the WMDR each station can report one or more observing schedules (per sensor) using the `data generation` entity. See the bottom right of the diagram at [https://wmo-im.github.io/wmdr/html/index.htm?goto=1](https://wmo-im.github.io/wmdr/html/index.htm?goto=1). Also copied in the screen shot below.

![image](https://user-images.githubusercontent.com/29403616/217241896-d8258549-95f4-49a9-8dfd-5de6af6188cc.png)

Under `sampling` we can specify both the `samplingTimePeriod` and the number of samples in that time period (`samplesPerTimePeriod`).

@joergklausen please correct me if I am wrong.
